File: sources/net.sf.j2s.core/test/src/com/falstad/circuit/MonostableElm.js
(function(){var P$=Clazz.newPackage("com.falstad.circuit"),I$=[];
var C$=Clazz.newClass(P$, "MonostableElm", null, 'com.falstad.circuit.ChipElm');

C$.$clinit$ = function() {Clazz.load(C$, 1);
}

Clazz.newMeth(C$, '$init0$', function () {
var c;if((c = C$.superclazz) && (c = c.$init0$))c.apply(this);
this.prevInputValue = false;
this.retriggerable = false;
this.triggered = false;
this.lastRisingEdge = 0;
this.delay = 0;
}, 1);

Clazz.newMeth(C$, '$init$', function () {
this.prevInputValue = false;
this.retriggerable = false;
this.triggered = false;
this.lastRisingEdge = 0;
this.delay = 0.01;
}, 1);

Clazz.newMeth(C$, 'c$$I$I', function (xx, yy) {
C$.superclazz.c$$I$I.apply(this, [xx, yy]);
C$.$init$.apply(this);
}, 1);

Clazz.newMeth(C$, 'c$$I$I$I$I$I$java_util_StringTokenizer', function (xa, ya, xb, yb, f, st) {
C$.superclazz.c$$I$I$I$I$I$java_util_StringTokenizer.apply(this, [xa, ya, xb, yb, f, st]);
C$.$init$.apply(this);
this.retriggerable =  Boolean.from(st.nextToken()).booleanValue();
this.delay =  new Double(st.nextToken()).doubleValue();
}, 1);

Clazz.newMeth(C$, 'getChipName', function () {
return "Monostable";
});

Clazz.newMeth(C$, 'setupPins', function () {
this.sizeX = 2;
this.sizeY = 2;
this.pins = Clazz.array((I$[0]||(I$[0]=Clazz.load(Clazz.load('com.falstad.circuit.ChipElm').Pin))), [this.getPostCount()]);
this.pins[0] = Clazz.new_((I$[0]||(I$[0]=Clazz.load(Clazz.load('com.falstad.circuit.ChipElm').Pin))).c$$I$I$S, [this, null, 0, 2, ""]);
this.pins[0].clock = true;
this.pins[1] = Clazz.new_((I$[0]||(I$[0]=Clazz.load(Clazz.load('com.falstad.circuit.ChipElm').Pin))).c$$I$I$S, [this, null, 0, 3, "Q"]);
this.pins[1].output = true;
this.pins[2] = Clazz.new_((I$[0]||(I$[0]=Clazz.load(Clazz.load('com.falstad.circuit.ChipElm').Pin))).c$$I$I$S, [this, null, 1, 3, "Q"]);
this.pins[2].output = true;
this.pins[2].lineOver = true;
});

Clazz.newMeth(C$, 'getPostCount', function () {
return 3;
});

Clazz.newMeth(C$, 'getVoltageSourceCount', function () {
return 2;
});

Clazz.newMeth(C$, 'execute', function () {
if (this.pins[0].value && this.prevInputValue != this.pins[0].value   && (this.retriggerable || !this.triggered ) ) {
this.lastRisingEdge = (I$[1]||(I$[1]=Clazz.load('com.falstad.circuit.CircuitElm'))).sim.t;
this.pins[1].value = true;
this.pins[2].value = false;
this.triggered = true;
}if (this.triggered && (I$[1]||(I$[1]=Clazz.load('com.falstad.circuit.CircuitElm'))).sim.t > this.lastRisingEdge + this.delay  ) {
this.pins[1].value = false;
this.pins[2].value = true;
this.triggered = false;
}this.prevInputValue = this.pins[0].value;
});

Clazz.newMeth(C$, 'dump', function () {
return C$.superclazz.prototype.dump.apply(this, []) + " " + this.retriggerable + " " + new Double(this.delay).toString() ;
});

Clazz.newMeth(C$, 'getDumpType', function () {
return 194;
});

Clazz.newMeth(C$, 'getEditInfo$I', function (n) {
if (n == 2) {
var ei = Clazz.new_((I$[2]||(I$[2]=Clazz.load('com.falstad.circuit.EditInfo'))).c$$S$D$D$D,["", 0, -1, -1]);
ei.checkbox = Clazz.new_((I$[3]||(I$[3]=Clazz.load('a2s.Checkbox'))).c$$S$Z,["Retriggerable", this.retriggerable]);
return ei;
}if (n == 3) {
var ei = Clazz.new_((I$[2]||(I$[2]=Clazz.load('com.falstad.circuit.EditInfo'))).c$$S$D$D$D,["Period (s)", this.delay, 0.001, 0.1]);
return ei;
}return C$.superclazz.prototype.getEditInfo$I.apply(this, [n]);
});

Clazz.newMeth(C$, 'setEditValue$I$com_falstad_circuit_EditInfo', function (n, ei) {
if (n == 2) {
this.retriggerable = ei.checkbox.getState();
}if (n == 3) {
this.delay = ei.value;
}C$.superclazz.prototype.setEditValue$I$com_falstad_circuit_EditInfo.apply(this, [n, ei]);
});

Clazz.newMeth(C$);
})();
//Created 2017-12-17 19:28:20

File: sources/net.sf.j2s.core/test/src/com/falstad/circuit/MonostableElm.java
package com.falstad.circuit;

import java.util.StringTokenizer;

import a2s.Checkbox;

    class MonostableElm extends ChipElm {

	//Used to detect rising edge
	private boolean prevInputValue=false;
	private boolean retriggerable=false;
	private boolean triggered=false;
	private double lastRisingEdge=0;
	private double delay=0.01;

	public MonostableElm(int xx, int yy) { super(xx, yy); }
	public MonostableElm(int xa, int ya, int xb, int yb, int f,
			    StringTokenizer st) {
	    super(xa, ya, xb, yb, f, st);
	retriggerable=new Boolean(st.nextToken()).booleanValue();
	delay=new Double(st.nextToken()).doubleValue();
	}
	String getChipName() { return "Monostable"; }
	void setupPins() {
	    sizeX = 2;
	    sizeY = 2;
	    pins = new Pin[getPostCount()];
	    pins[0] = new Pin(0, SIDE_W, "");
	    pins[0].clock = true;
	    pins[1] = new Pin(0, SIDE_E, "Q");
	    pins[1].output=true;
	    pins[2] = new Pin(1, SIDE_E, "Q");
	    pins[2].output=true;
	    pins[2].lineOver=true;
	}
	int getPostCount() {
	return 3;
	}
	int getVoltageSourceCount() { return 2; }

	void execute() {

			if(pins[0].value&&prevInputValue!=pins[0].value&&(retriggerable||!triggered)){
			lastRisingEdge=sim.t;
			pins[1].value=true;
			pins[2].value=false;
			triggered=true;
			}

			if(triggered&&sim.t>lastRisingEdge+delay)
			{
			pins[1].value=false;
			pins[2].value=true;
			triggered=false;
			}
		prevInputValue=pins[0].value;
	   	}
	String dump(){
	   return super.dump() + " " + retriggerable + " " + delay;
	}
	int getDumpType() { return 194; }
	public EditInfo getEditInfo(int n) {
	    if (n == 2) {
		EditInfo ei = new EditInfo("", 0, -1, -1);
		ei.checkbox=new Checkbox("Retriggerable",retriggerable);
		return ei;
	    }
	    if (n == 3) {
		EditInfo ei = new EditInfo("Period (s)",delay, 0.001,0.1);
		return ei;
	    }
	    return super.getEditInfo(n);
	}
	public void setEditValue(int n, EditInfo ei) {
	    if (n == 2) {
		retriggerable=ei.checkbox.getState();
	    }
	    if (n == 3) {
		delay=ei.value;
	    }
	    super.setEditValue(n, ei);
	}
    }
